### ISSUE TITLE
AQL traversal path issues

### ISSUE THREAD

**Issue by mgm81** [arangodb/arangodb#2853]
### this is an AQL-related issue:
[x] I'm using graph features

I am writing a graph traversal query in AQL (3.1.22, windows 7) where the start vertex is from a filtered vertex collection. For some of the paths (of some of the start vertices) that are returned, I get a start vertex which is not present on the path. The edges and the remaining vertices on the path are as expected.

However, if I hard code the correct start vertex (for a path that had an incorrect start vertex in the vertex list), I get the correct path object. Is this due to incorrect optimization in AQL and if so, how can I turn off optimization?

**Comment by dothebart** [comment]
Hi, 
can you provide a minimal dump reproducing this issue?

**Comment by mgm81** [comment]
Sorry I am new to this, can you please explain how to create a minimal dump? 

Here is the query that gives me the offending path object (shown here for ono==27. It is not limited to edges with only this attribute, but conversely does not happen for all paths). The start vertex returned is not associated with the edge.

`Let path = (For v in @@cname 
                filter (v.net == "PSN")
                FOR v, e, p IN 1..2 ANY v GRAPH @gname FILTER p.edges[*].ron ALL == @nOp && p.edges[*].dir ALL == @dir && p.edges[*].ev ALL == @ev && p.edges[*].ete ALL == @eT && (p.vertices[2] != null? p.edges[0].ono == p.edges[1].ono: true)
                    RETURN p)
For p in path
    FILTER p.edges[0].ono == 27
return p

_Here is the result_

{	"edges": [{
        "_key": "133085153457",
        "_id": "mddaccess_baseEdge/133085153457",
        "_from": "mddaccess_baseNode/**133085153409**",
        "_to": "mddaccess_baseNode/133085153452",
        "_rev": "_VTeyOQy---",
        "ron": "A3241",
        "ete": "specp",
        "name": "PSN",
        "dir": "AB90985",
        "fName": "RDTMA",
        "ev": 1,
        "vn": 27,
        "ono": 27
      }
    ],
    "vertices": [
      {
        "_key": "133085153670",
        "_id": "mddaccess_baseNode/**133085153670**",
        "_rev": "_VTeyOcC---",
        "name": "Fopsea",
        "net": "PSN"
      },
      {
        "_key": "133085153452",
        "_id": "mddaccess_baseNode/133085153452",
        "_rev": "_VTeyOQq---",
        "name": "NA",
        "net": "MTD"
      }
    ]
  }
`

If I give the query the correct start vertex, I get the correct path object.

`Let path = (FOR v, e, p IN 1..2 ANY 'mddaccess_baseNode/133085153409' GRAPH @gname FILTER p.edges[*].ron ALL == @nOp && p.edges[*].dir ALL == @dir && p.edges[*].ev ALL == @ev && p.edges[*].ete ALL == @eT && (p.vertices[2] != null? p.edges[0].ono == p.edges[1].ono: true)
                    RETURN p)

For p in path
    FILTER p.edges[0].ono == 27
return p


{
    "edges": [
      {
        "_key": "133085153457",
        "_id": "mddaccess_baseEdge/133085153457",
        "_from": "mddaccess_baseNode/**133085153409**",
        "_to": "mddaccess_baseNode/133085153452",
        "_rev": "_VTeyOQy---",
        "ron": "A3241",
        "ete": "specp",
        "name": "PSN",
        "dir": "AB90985",
        "fName": "RDTMA",
        "ev": 1,
        "vn": 27,
        "ono": 27
      }
    ],
    "vertices": [
      {
        "_key": "133085153409",
        "_id": "mddaccess_baseNode/**133085153409**",
        "_rev": "_VTeyOOq---",
        "name": "Indic set",
        "net": "PSN"
      },
      {
        "_key": "133085153452",
        "_id": "mddaccess_baseNode/133085153452",
        "_rev": "_VTeyOQq---",
        "name": "NA",
        "net": "MTD"
      }
    ]
  }`

**Comment by dothebart** [comment]
Sorry for the late reply, somehow this got lost. 

I've tried this with the [knows_graph](https://docs.arangodb.com/3.2/Manual/Graphs/#the-knowsgraph)

which seems to be working properly:

```
db._query(`for pr IN persons FILTER pr.name == "Dave" FOR v,e,p IN 1..2 ANY pr GRAPH "knows_graph" FILTER e.vertex == "bob" RETURN p`).toArray()
[ 
  { 
    "edges" : [ 
      { 
        "_key" : "5489", 
        "_id" : "knows/5489", 
        "_from" : "persons/bob", 
        "_to" : "persons/dave", 
        "_rev" : "_VeoJ-fW--A", 
        "vertex" : "bob" 
      } 
    ], 
    "vertices" : [ 
      { 
        "_key" : "dave", 
        "_id" : "persons/dave", 
        "_rev" : "_VeoJ-fS--B", 
        "name" : "Dave" 
      }, 
      { 
        "_key" : "bob", 
        "_id" : "persons/bob", 
        "_rev" : "_VeoJ-fS--_", 
        "name" : "Bob" 
      } 
    ] 
  }, 
  { 
    "edges" : [ 
      { 
        "_key" : "5489", 
        "_id" : "knows/5489", 
        "_from" : "persons/bob", 
        "_to" : "persons/dave", 
        "_rev" : "_VeoJ-fW--A", 
        "vertex" : "bob" 
      }, 
      { 
        "_key" : "5486", 
        "_id" : "knows/5486", 
        "_from" : "persons/bob", 
        "_to" : "persons/charlie", 
        "_rev" : "_VeoJ-fW--_", 
        "vertex" : "bob" 
      } 
    ], 
    "vertices" : [ 
      { 
        "_key" : "dave", 
        "_id" : "persons/dave", 
        "_rev" : "_VeoJ-fS--B", 
        "name" : "Dave" 
      }, 
      { 
        "_key" : "bob", 
        "_id" : "persons/bob", 
        "_rev" : "_VeoJ-fS--_", 
        "name" : "Bob" 
      }, 
      { 
        "_key" : "charlie", 
        "_id" : "persons/charlie", 
        "_rev" : "_VeoJ-fS--A", 
        "name" : "Charlie" 
      } 
    ] 
  } 
]
```
If you send me the [dump](https://docs.arangodb.com/3.2/Manual/Administration/Arangodump.html) of your Graph I can have a deeper look.

Usually its a good idea to deconstruct the query into its bits. 
I.e. does the outer `FOR` and its `FILTER` statement realy return the same start-Vertex as you specify by the string? 
One could try whether using `v._id` instead of `v` as a starting point of the graph.

**Comment by mgm81** [comment]
Thanks for your reply. I went with the custom traversal objects and that seems to be working correctly. I am not sure what it is in the dataset that causes only that one path out of 100 or so paths to return the wrong starting vertex in the path. This is shown in my previous reply (node keys are in bold in the result set). 

What I don't quite get is that even if my query is returning the wrong start vertex, shouldn't the path object returned contain edges and vertices that are connected (i.e. vertex _ids in the path should match _from and _to of the edge in the path)?

I would have done the dump, but there is some proprietary data that prevents me from doing this. I think I will close the issue for now and re-open/create the issue if I see this again in the future.

**Comment by jsteemann** [comment]
If you are using 3.1.22 as reported above, an upgrade to 3.1.27 (latest 3.1) may help. I remember there were some traversal issues fixed in 3.1.27 and 3.1.24.

**Comment by dothebart** [comment]
Closing as solved.
